api.fbs

```diff
@@ -0,0 +1,140 @@
+namespace proto;
+
+table MessageHeader {
+  key   : string  (id: 0);
+  value : [ubyte] (id: 1);
+}
+
+// Message is the core immutable unit that is propagated/stored.
+table Message {
+  value   : [ubyte]         (id: 0);
+  key     : [ubyte]         (id: 1); // Message key used for log compaction
+  reply   : string          (id: 2); // NATS subject to reply to
+  headers : [MessageHeader] (id: 3); // Message headers
+  subject : string          (id: 4); // NATS subject message was received on
+  offset  : int64           (id: 5); // Monotonic message offset in the stream
+}
+
+table CreateStreamRequest {
+  subject           : string (id: 0); // NATS subject to attach to
+  stream            : string (id: 1); // Stream name (unique per subject)
+  group             : string (id: 2); // Partitions NATS subject amongst group members
+  replicationFactor : int32  (id: 3); // Number of stream replicas (-1 to replicate to the entire cluster)
+  partitions        : uint32 (id: 4); // Number of stream partitions
+}
+
+table CreateStreamResponse {
+}
+
+// AckPolicy controls the behavior of message acknowledgements.
+enum AckPolicy : uint8 {
+  LEADER = 0, // The ack will be sent once the leader has written the message to its log
+  ALL    = 1, // The ack will be sent after the ISR replicas have written the message to their logs
+  NONE   = 2, // No ack will be sent
+}
+
+// PublishRequest is used for both gRPC Publish requests as well as direct NATS
+// publishes (see the README for the envelope format).
+table PublishRequest {
+  message       : [ubyte]      (id: 0, nested_flatbuffer: "Message");
+  correlationId : [ubyte]      (id: 1); // User-supplied id to correlate acks to publishes
+  ackInbox      : string       (id: 2); // NATS subject to publish acks to
+  ackPolicy     : AckPolicy    (id: 3); // Controls the behavior of acks
+}
+
+table PublishResponse {
+  ack : Ack (id: 0); // The ack for the published message if AckPolicy was not NONE
+}
+
+// Ack represents an acknowledgement that a message was committed to a stream partition.
+table Ack {
+  correlationId    : [ubyte]   (id: 0); // User-supplied value from the message
+  stream           : string    (id: 1); // Name of the stream
+  partitionSubject : string    (id: 2); // NATS subject the partition is attached to
+  msgSubject       : string    (id: 3); // NATS subject the message was received on
+  offset           : int64     (id: 4); // Stream offset the message was committed to
+  ackInbox         : string    (id: 5); // NATS subject to publish acks to
+  ackPolicy        : AckPolicy (id: 6); // The AckPolicy sent on the message
+}
+
+enum StartPosition : uint8 {
+  NEW_ONLY  = 0, // Start at new messages after the latest
+  OFFSET    = 1, // Start at a specified offset
+  EARLIEST  = 2, // Start at the oldest message
+  LATEST    = 3, // Start at the newest message
+  TIMESTAMP = 4, // Start at a specified (ack) timestamp
+}
```

Comment:
Imo start position could be an Union instead. This should make consistency easier (eg: not setting an offset when start position type is `offset`)